File: app/src/main/java/com/coryswainston/smart/dictionary/services/DetectorProcessor.java
package com.coryswainston.smart.dictionary.services;

import android.graphics.Rect;
import android.util.SparseArray;

import com.google.android.gms.vision.Detector;
import com.google.android.gms.vision.text.Text;
import com.google.android.gms.vision.text.TextBlock;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Processes detected text
 */

public class DetectorProcessor implements Detector.Processor<TextBlock> {

    private Callback callback;

    @Override
    public void release() {
        // empty
    }

    @Override
    public void receiveDetections(Detector.Detections<TextBlock> detections) {
        SparseArray<TextBlock> textBlocks = detections.getDetectedItems();

        List<Text> wordBlocks = new ArrayList<>();

        Map<String, Rect> words = new HashMap<>();
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < textBlocks.size(); i++) {
            sb.append(textBlocks.valueAt(i).getValue());
            List<? extends Text> lines = textBlocks.valueAt(i).getComponents();
            for (int j = 0; j < lines.size(); j++) {
                List<? extends Text> wordList = lines.get(j).getComponents();
                for (int k = 0; k < wordList.size(); k++) {
                    words.put(wordList.get(k).getValue(), wordList.get(k).getBoundingBox());
                    wordBlocks.add(wordList.get(k));
                }
            }
        }

        callback.callback(wordBlocks, sb.toString());
    }

    public DetectorProcessor withCallback(Callback callback) {
        this.callback = callback;
        return this;
    }

    public interface Callback {
        void callback(List<Text> blocks, String detectedText);
    }
}

File: app/src/main/java/com/coryswainston/smart/dictionary/services/DetectorProcessor.kt
package com.coryswainston.smart.dictionary.services

import android.graphics.Rect
import com.google.android.gms.vision.Detector
import com.google.android.gms.vision.Detector.Detections
import com.google.android.gms.vision.text.Text
import com.google.android.gms.vision.text.TextBlock
import java.util.*

/**
 * Processes detected text
 */
class DetectorProcessor : Detector.Processor<TextBlock> {
    private lateinit var callback: (List<Text>?, String?) -> Unit

    override fun release() {
        // empty
    }

    override fun receiveDetections(detections: Detections<TextBlock>) {
        val textBlocks = detections.detectedItems
        val wordBlocks: MutableList<Text> = ArrayList()
        val words: MutableMap<String, Rect> = HashMap()
        val sb = StringBuilder()
        for (i in 0 until textBlocks.size()) {
            sb.append(textBlocks.valueAt(i).value)
            val lines = textBlocks.valueAt(i).components
            for (j in lines.indices) {
                val wordList = lines[j].components
                for (k in wordList.indices) {
                    words[wordList[k].value] = wordList[k].boundingBox
                    wordBlocks.add(wordList[k])
                }
            }
        }
        callback(wordBlocks, sb.toString())
    }

    fun withCallback(callback: (List<Text>?, String?) -> Unit): DetectorProcessor {
        this.callback = callback
        return this
    }
}